burn-tensor/src/tensor/api/base.rs

```diff
@@ -168,6 +168,43 @@ where
     }
 }
 
+impl<B> Tensor<B, 2, Int>
+where
+    B: Backend,
+{
+    fn to_nested_vec(&self) -> Vec<Vec<B::IntElem>> {
+        let data = self.to_data();
+        let mut result = vec![vec![B::IntElem::default(); self.dims()[1]]; self.dims()[0]];
+        for (i, val) in data.value.iter().enumerate() {
+            let row = i / self.dims()[1];
+            let col = i % self.dims()[1];
+            result[row][col] = *val;
+        }
+        result
```

Comment:
This could be made as a separate standalone function that can take the multi dimensional tensor and return formatted string string. This way you don't need to worry about nested vectors. 

The challenging part is to convert the flat data into nested array. But I am sure it's doing. Here is an example output how numpy does:


```
>>> np.ones([15, 15,15, 15])
array([[[[1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         ...,
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.]],

        [[1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         ...,
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.]],

        [[1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         ...,
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.]],

        ...,

        [[1., 1., 1., ..., 1., 1., 1.],
         [1., 1., 1., ..., 1., 1., 1.],

```

Reply:
My current strategy is to try and recursively format the tensor as a 2D vector for each dimension, indenting based on the dimensionality level. It's still a work in progress because I need to figure out how to properly slice up tensors.

Reply:
@agelas @nathanielsimard 

I came across this [project serde-dim](https://github.com/RReverser/serde-ndim) that looks interesting. Maybe there is a something you can learn from it: https://github.com/RReverser/serde-ndim/blob/main/src/ser.rs . What this project is trying to accomplish is basically convert one dimensional data array into multi dimensional array. Pretty what you're trying to accomplish.

Reply:
@antimora Looks interesting, I'll give it a look. But your description seems almost the opposite of what this PR is for, ie we're going from multi-dimensional array -> "1-dimensional" string representation and not the other way around.

Reply:
> @antimora Looks interesting, I'll give it a look. But your description seems almost the opposite of what this PR is for, ie we're going from multi-dimensional array -> "1-dimensional" string representation and not the other way around.

The tensor data is stored as one-dimensional vector and shape contains the dimensions so you know how to decompose into rows and columns. The link I shared basically is trying to take the serialized data (1d) and deserialize into n-dimensional array by recursion. By coincidence I came across this code and I thought there might be something to learn. I haven't dived deep to see if this direct helps you but in principle by its description it should. I hope it does not confuse you =)